Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs

```diff
@@ -2233,24 +2240,55 @@ private bool PrepareInContainerPaths()
             return true;
         }
 
-        private bool PrepareDirectoryOutputs()
+        private async Task<bool> PrepareDirectoryOutputs(HashSet<AbsolutePath> preserveOutputWhitelist)
         {
             foreach (var directoryOutput in m_pip.DirectoryOutputs)
             {
                 try
                 {
-                    if (!directoryOutput.IsSharedOpaque)
+                    string directoryPathStr = directoryOutput.Path.ToString(m_pathTable);
+                    bool dirExist = FileUtilities.DirectoryExistsNoFollow(directoryPathStr);
+
+                    if (directoryOutput.IsSharedOpaque)
                     {
-                        PreparePathForOutputDirectory(directoryOutput.Path, createIfNonExistent: true);
+                        // Ensure it exists.
+                        if (!dirExist)
+                        {
+                            FileUtilities.CreateDirectory(directoryPathStr);
+                        }
                     }
                     else
                     {
-                        // Ensure it exists.
-                        string directoryPath = directoryOutput.Path.ToString(m_pathTable);
-
-                        if (!FileUtilities.DirectoryExistsNoFollow(directoryPath))
+                        if (dirExist && ShouldPreserveOutput(directoryOutput.Path, preserveOutputWhitelist))
                         {
-                            FileUtilities.CreateDirectory(directoryPath);
+                            using (var wrapper = Pools.GetFileArtifactList())
+                            {
+                                var files = wrapper.Instance;
+                                FileUtilities.EnumerateDirectoryEntries(
+                                    directoryPathStr,
+                                    recursive: true,
+                                    handleEntry: (currentDir, name, attributes) =>
+                                    {
+                                        if ((attributes & FileAttributes.Directory) == 0)
+                                        {
+                                            var path = AbsolutePath.Create(m_pathTable, currentDir).Combine(m_pathTable, name);
+                                            var fileArtifact = FileArtifact.CreateOutputFile(path);

```

Comment:
It looks like the implementation of m_makeOutputPrivate immediate switches the FileArtifact to a string path. Can you just update that function signature and remove the translation?

Reply:
nice catch, it was pretty subtle. converted the parameter to string. 

---
In reply to: [287419128](https://github.com/microsoft/BuildXL/pull/369#discussion_r287419128) [](ancestors = 287419128)